source/agora/common/Transaction.d

```diff
@@ -150,3 +151,112 @@ public struct Input
         hashPart(this.index, dg);
     }
 }
+
+/*******************************************************************************
+
+    Checks whether the transaction is coinbase
+
+    Params:
+        tx = Type of `Transaction`
+
+    Return:
+        Return true if this transaction is a coinbase
+
+*******************************************************************************/
+
+bool isCoinbaseTx (Transaction tx)
+{
+    return tx.inputs.length == 1 && tx.inputs[0] == Input.init;
+}
+
+/*******************************************************************************
+
+    Creates a new coinbase transaction
+
+    Params:
+        address = The public key that can redeem this output
+        value = The initial value
+
+    Return:
+        Return coinbase transation
+
+*******************************************************************************/
+
+Transaction newCoinbaseTX (PublicKey address, Amount value = 0)
+{
+    return Transaction(
+        [Input(Hash.init, 0)],
+        [Output(value, address)]
+    );
+}
+
+/*******************************************************************************
+
+    Get sum of `Input`
+
+    Params:
+        tx = `Transaction`
+        findOutput = delegate for finding `Output`
+
+    Return:
+        Sum of `Input` in the `Transaction`
+
+*******************************************************************************/
+long getSumInput (Transaction tx, Output* delegate (Hash hash, uint index) findOutput)
+{
+    return tx.inputs.map!(a => findOutput(a.previous, a.index)).filter!(a => a !is null).map!(a => a.value).sum();
+}
+
+/*******************************************************************************
+
+    Get sum of `Output`
+
+    Params:
+        tx = `Transaction`
+
+    Return:
+        Sum of `Output` in the `Transaction`
+
+*******************************************************************************/
+
+long getSumOutput (Transaction tx)
+{
+    return tx.outputs.map!(a => a.value).sum();
+}
+
+/*******************************************************************************
+
+    Get result of transaction verification
+
+    Params:
+        tx = `Transaction`
+        findOutput = delegate for finding `Output`
+
+    Return:
+        Return true if this transaction is verified.
+
+*******************************************************************************/
+
+public bool verifyData (Transaction tx, Output* delegate (Hash hash, uint index) findOutput)
+{
+    if (tx.isCoinbaseTx())
+    {
+        if (tx.outputs.length == 0)
+            return false;
+        else
+            return true;
+    }
+    else
+    {
+        if (tx.inputs.length == 0)
+            return false;
+
+        if (tx.outputs.length == 0)
+            return false;
```

Comment:
Move this `if` before `if (tx.isCoinbaseTx())` ?

Reply:
Modified